### PR TITLE
KAFKA-17606: Include Rat errors in GitHub workflow summary

### PR DESCRIPTION
**Pull request by LoganZhuZzz**
In the current GitHub workflow, it has become difficult to locate Rat errors when running `check -x test`. To improve readability and make it easier to identify Rat errors, we need to include them in the workflow summary.

## Key Modifications
- Added a call to `rat.py` in the `build.yml` file to annotate Rat errors after the `Compile and validate` step.
- Ensured that the annotation for Rat reports will execute **only when the `Compile and validate` step fails.** 

## Purpose
- Enhance the usability of the GitHub workflow, allowing developers to more easily identify and fix Rat errors.
- Improve workflow efficiency by reducing the time spent manually searching for errors.

## Notes
- **Please ensure that all relevant workflow tests are run before merging this PR to confirm that the new feature operates correctly.**

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by LoganZhuZzz** [comment]
@chia7712 please take a look,thanks!

File: .github/workflows/build.yml [hunk 1]
```diff
@@ -68,6 +68,11 @@ jobs:
           ./gradlew --build-cache --info  \
           ${{ inputs.is-public-fork && '--no-scan' || '--scan' }} \
           check -x test
+      - name: Annotate Rat errors
```
Comment (mumrah) [inline .github/workflows/build.yml]: Can you move this down with the checkstyle step?

File: .github/workflows/build.yml [hunk 2]
```diff
@@ -68,6 +68,11 @@ jobs:
...
+        if: ${{ failure() }}
```
Comment (mumrah) [inline .github/workflows/build.yml]: We should only run this on java 21 to avoid duplicate reports from the parallel builds

**Review by mumrah — COMMENTED** [review]
Thanks for the patch @LoganZhuZzz! 

Left a few minor comments inline. Can you temporarily add a couple of files without a license so we can see the script working in this PR?

**Comment by LoganZhuZzz** [comment]
> Thanks for the patch @LoganZhuZzz!
> 
> Left a few minor comments inline. Can you temporarily add a couple of files without a license so we can see the script working in this PR?

Thank you for the feedback on the patch @mumrah 
I have made the following changes:

1. Temporarily added test files without licenses to validate the unapproved licenses summary in the GitHub workflow.
2. Moved the `Annotate Rat errors` step below the `Annotate checkstyle errors` step for better clarity.
3. Added a version check to ensure annotation steps run only on Java 21 to prevent duplicate reports.

Please let me know if there’s anything else you would like me to address!

**Comment by LoganZhuZzz** [comment]
This pull request successfully integrates the annotation of Rat errors into the GitHub workflow summary.

**Comment by mumrah** [comment]
Thanks @LoganZhuZzz, the script output looks good. Can you try the following for me:

1) Add a `ignoreRatFailure` parameter to build.gradle that controls `rat { failOnError }`
2) Pass in this param in build.yml when we invoke `check -x test`
3) Modify rat.py to exit 1 if it finds any problems

This way, the link to the failed workflow will take us to the Rat output rather than the first failing step.

Also, could you include "::notice" output in the script similar to this https://github.com/apache/kafka/blob/trunk/.github/scripts/checkstyle.py#L64? This is how we actually create the annotations. Note that the line number starts with 1. https://docs.github.com/en/actions/writing-workflows/choosing-what-your-workflow-does/workflow-commands-for-github-actions#setting-a-notice-message

**Comment by LoganZhuZzz** [comment]
> Thanks @LoganZhuZzz, the script output looks good. Can you try the following for me:
> 
> 1. Add a `ignoreRatFailure` parameter to build.gradle that controls `rat { failOnError }`
> 2. Pass in this param in build.yml when we invoke `check -x test`
> 3. Modify rat.py to exit 1 if it finds any problems
> 
> This way, the link to the failed workflow will take us to the Rat output rather than the first failing step.
> 
> Also, could you include "::notice" output in the script similar to this https://github.com/apache/kafka/blob/trunk/.github/scripts/checkstyle.py#L64? This is how we actually create the annotations. Note that the line number starts with 1. https://docs.github.com/en/actions/writing-workflows/choosing-what-your-workflow-does/workflow-commands-for-github-actions#setting-a-notice-message

Thanks for the guidance. @mumrah 
I have fixed the above issues, If you notice any areas for improvement or any bad practices, please let me know.

File: build.gradle [hunk 3]
```diff
@@ -210,6 +210,10 @@ def determineCommitId() {
 apply from: file('wrapper.gradle')
 
 if (repo != null) {
+  ext {
+    ignoreRatFailure = project.hasProperty('ignoreRatFailure') ? project.ignoreRatFailure.toBoolean() : false
```
Comment (mumrah) [inline build.gradle]: Can you relocate this with other property handling? (Along with userMaxForks, etc)

File: .github/workflows/build.yml [hunk 4]
```diff
@@ -67,7 +67,7 @@ jobs:
         run: |
           ./gradlew --build-cache --info  \
           ${{ inputs.is-public-fork && '--no-scan' || '--scan' }} \
-          check -x test
+          check -x test  -PignoreRatFailure=true
```
Comment (mumrah) [inline .github/workflows/build.yml]: Can you move the property before the task on a separate line?

File: .github/scripts/rat.py [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+    else:
+        print(f"{total_approved_licenses} approved licenses")
+        print(f"{total_unapproved_licenses} unapproved licenses")
+
+        print("Files with unapproved licenses:")
+        for file in all_unapproved_files:
+            rel_path = os.path.relpath(file, workspace_path)
+            print(f"::notice file={rel_path},title=Unapproved License::File with unapproved license")
+
+    exit(1 if total_unapproved_licenses > 0 else 0)
```
Comment (mumrah) [inline .github/scripts/rat.py]: Can you make this a regular if/else rather than inlined?

**Review by mumrah — COMMENTED** [review]
Thanks for the updates @LoganZhuZzz! Just a few small comments. 

Can you try this out on your fork and include a link to the results?

**Comment by LoganZhuZzz** [comment]
Hi David @mumrah ! Here are the results on my fork:

### Compile and Validate Results
b8d38b77eb24bd5773c4949083d5e2c

### Annotate Rat Errors Results
09c6fe160eb0abde9296ce9920b0fb3

[Link to Results](https://github.com/LoganZhuZzz/kafka/actions/runs/11091913918/job/30816286950)

---

### Questions

1. After modifying the `Annotate Rat errors` step from `if: ${{ failure() && matrix.java == '21' }}` to `if: ${{ matrix.java == '21' }}`, the `Annotate Rat errors` step was able to execute. Is this behavior in line with our expectations?
2. Do the results from `Annotate Rat errors` meet our expectations?

### Root Cause of `Annotate Rat Errors` Step Not Executing
The reason for this issue is that in the `Compile and validate` phase, we execute:

```bash
./gradlew --build-cache --info \
${{ inputs.  is-public-fork && '--no-scan' || '--scan' }} \
-PignoreRatFailure=true \
check -x test
```

When the `ignoreRatFailure` parameter is set to true, the `failOnError` of the `Gradle RAT` task is set to false.This causes the `failure()` condition in the `Annotate Rat errors` step of the workflow (`build.yml`) to evaluate as `false`, preventing the `Annotate Rat errors` step from being executed.   Does this align with what we expect?

**Comment by mumrah** [comment]
> When the ignoreRatFailure parameter is set to true, the failOnError of the Gradle RAT task is set to false.This causes the failure() condition in the Annotate Rat errors step of the workflow (build.yml) to evaluate as false, preventing the Annotate Rat errors step from being executed. Does this align with what we expect?

I see what you mean. In this case, let's do the same as we do for checkstyle and let RAT failures fail the build. Then, we will hit the `failure()` condition as expected. To do this, I think we just need to remove the `ignoreRatFailure` property.

**Comment by LoganZhuZzz** [comment]
Now, it meets our expectations. Any suggestions? @mumrah 

## Output Results
eb639b877fc14ba94f172149b6e13ce
edd5a162cc733152aad8930c8a35890

[Link to Results](https://github.com/apache/kafka/actions/runs/11098437481/job/30831116931?pr=17280)

**Comment by mumrah** [comment]
@LoganZhuZzz, looks great! Please revert the temporary files and I'll merge this.

**Comment by LoganZhuZzz** [comment]
> @LoganZhuZzz, looks great! Please revert the temporary files and I'll merge this.

Thanks for the guidance. @mumrah

**Review by mumrah — APPROVED** [review]
LGTM, thanks @LoganZhuZzz!

edit: Also, thanks for the well written and well formatted PR :)

**Pull request merged**